File: nodejs/src/controllers/PersonasController.ts
import { Request, Response } from 'express';
import { IPersonaService } from '../services/IPersonaService';
import { PersonaSearchDto } from '../dto/PersonaDto';
import { AuthRequest } from '../middleware/auth';

export class PersonasController {
  constructor(private personaService: IPersonaService) {}

  async getAll(req: AuthRequest, res: Response): Promise<void> {
    try {
      const personas = await this.personaService.findAll();
      res.json(personas);
    } catch (error) {
      console.error('Error al obtener todas las personas:', error);
      res.status(500).json({ message: 'Error interno del servidor' });
    }
  }

  async getById(req: AuthRequest, res: Response): Promise<void> {
    try {
      const id = parseInt(req.params.id);
      const persona = await this.personaService.findById(id);
      
      if (!persona) {
        res.status(404).json({ message: `No se encontró la persona con ID: ${id}` });
        return;
      }
      
      res.json(persona);
    } catch (error) {
      console.error('Error al obtener persona:', error);
      res.status(500).json({ message: 'Error interno del servidor' });
    }
  }

  async create(req: AuthRequest, res: Response): Promise<void> {
    try {
      const persona = await this.personaService.create(req.body);
      res.status(201).json(persona);
    } catch (error: any) {
      console.error('Error al crear persona:', error);
      
      if (error.message.includes('Ya existe')) {
        res.status(400).json({ message: error.message });
        return;
      }
      
      res.status(500).json({ message: 'Error interno del servidor' });
    }
  }

  async update(req: AuthRequest, res: Response): Promise<void> {
    try {
      const id = parseInt(req.params.id);
      const persona = await this.personaService.update(id, req.body);
      res.json(persona);
    } catch (error: any) {
      console.error('Error al actualizar persona:', error);
      
      if (error.message.includes('No se encontró')) {
        res.status(404).json({ message: error.message });
        return;
      }
      
      if (error.message.includes('Ya existe')) {
        res.status(400).json({ message: error.message });
        return;
      }
      
      res.status(500).json({ message: 'Error interno del servidor' });
    }
  }

  async delete(req: AuthRequest, res: Response): Promise<void> {
    try {
      const id = parseInt(req.params.id);
      await this.personaService.delete(id);
      res.status(204).send();
    } catch (error: any) {
      console.error('Error al eliminar persona:', error);
      
      if (error.message.includes('No se encontró')) {
        res.status(404).json({ message: error.message });
        return;
      }
      
      res.status(500).json({ message: 'Error interno del servidor' });
    }
  }

  async search(req: AuthRequest, res: Response): Promise<void> {
    try {
      const searchDto: PersonaSearchDto = req.body;
      const personas = await this.personaService.search(searchDto);
      res.json(personas);
    } catch (error) {
      console.error('Error al buscar personas:', error);
      res.status(500).json({ message: 'Error interno del servidor' });
    }
  }

  async validateIdentificacion(req: AuthRequest, res: Response): Promise<void> {
    try {
      const { identificacion } = req.params;
      const excludeId = req.query.excludeId ? parseInt(req.query.excludeId as string) : undefined;
      
      const isValid = await this.personaService.validateIdentificacion(identificacion, excludeId);
      res.json({ isValid });
    } catch (error) {
      console.error('Error al validar identificación:', error);
      res.status(500).json({ message: 'Error interno del servidor' });
    }
  }
}

File: dotnet/PersonasABM.API/Controllers/PersonasController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using PersonasABM.Application.DTOs;
using PersonasABM.Application.Interfaces;

namespace PersonasABM.API.Controllers;

[ApiController]
[Route("api/[controller]")]
[Authorize]
public class PersonasController : ControllerBase
{
    private readonly IPersonaService _personaService;
    private readonly ILogger<PersonasController> _logger;

    public PersonasController(IPersonaService personaService, ILogger<PersonasController> logger)
    {
        _personaService = personaService;
        _logger = logger;
    }

    /// <summary>
    /// Obtiene todas las personas
    /// </summary>
    [HttpGet]
    public async Task<ActionResult<IEnumerable<PersonaDto>>> GetAll()
    {
        try
        {
            var personas = await _personaService.GetAllAsync();
            return Ok(personas);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error al obtener todas las personas");
            return StatusCode(500, "Error interno del servidor");
        }
    }

    /// <summary>
    /// Obtiene una persona por ID
    /// </summary>
    [HttpGet("{id}")]
    public async Task<ActionResult<PersonaDto>> GetById(int id)
    {
        try
        {
            var persona = await _personaService.GetByIdAsync(id);
            if (persona == null)
            {
                return NotFound($"No se encontró la persona con ID: {id}");
            }
            return Ok(persona);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error al obtener persona con ID: {Id}", id);
            return StatusCode(500, "Error interno del servidor");
        }
    }

    /// <summary>
    /// Crea una nueva persona (Solo Admin)
    /// </summary>
    [HttpPost]
    [Authorize(Roles = "Admin")]
    public async Task<ActionResult<PersonaDto>> Create([FromBody] CreatePersonaDto createPersonaDto)
    {
        try
        {
            if (!ModelState.IsValid)
            {
                return BadRequest(ModelState);
            }

            var persona = await _personaService.CreateAsync(createPersonaDto);
            return CreatedAtAction(nameof(GetById), new { id = persona.Id }, persona);
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(ex.Message);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error al crear persona");
            return StatusCode(500, "Error interno del servidor");
        }
    }

    /// <summary>
    /// Actualiza una persona existente (Solo Admin)
    /// </summary>
    [HttpPut("{id}")]
    [Authorize(Roles = "Admin")]
    public async Task<ActionResult<PersonaDto>> Update(int id, [FromBody] UpdatePersonaDto updatePersonaDto)
    {
        try
        {
            if (!ModelState.IsValid)
            {
                return BadRequest(ModelState);
            }

            var persona = await _personaService.UpdateAsync(id, updatePersonaDto);
            return Ok(persona);
        }
        catch (KeyNotFoundException ex)
        {
            return NotFound(ex.Message);
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(ex.Message);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error al actualizar persona con ID: {Id}", id);
            return StatusCode(500, "Error interno del servidor");
        }
    }

    /// <summary>
    /// Elimina una persona (Solo Admin)
    /// </summary>
    [HttpDelete("{id}")]
    [Authorize(Roles = "Admin")]
    public async Task<IActionResult> Delete(int id)
    {
        try
        {
            await _personaService.DeleteAsync(id);
            return NoContent();
        }
        catch (KeyNotFoundException ex)
        {
            return NotFound(ex.Message);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error al eliminar persona con ID: {Id}", id);
            return StatusCode(500, "Error interno del servidor");
        }
    }

    /// <summary>
    /// Busca personas con filtros
    /// </summary>
    [HttpPost("search")]
    public async Task<ActionResult<IEnumerable<PersonaDto>>> Search([FromBody] PersonaSearchDto searchDto)
    {
        try
        {
            var personas = await _personaService.SearchAsync(searchDto);
            return Ok(personas);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error al buscar personas");
            return StatusCode(500, "Error interno del servidor");
        }
    }

    /// <summary>
    /// Valida si una identificación está disponible
    /// </summary>
    [HttpGet("validate-identificacion/{identificacion}")]
    public async Task<ActionResult<bool>> ValidateIdentificacion(string identificacion, [FromQuery] int? excludeId = null)
    {
        try
        {
            var isValid = await _personaService.ValidateIdentificacionAsync(identificacion, excludeId);
            return Ok(new { isValid });
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error al validar identificación: {Identificacion}", identificacion);
            return StatusCode(500, "Error interno del servidor");
        }
    }
}
